### DISCUSSION TITLE
关于aptos_framework::managed_coin & coin包的一些疑问

### DISCUSSION THREAD

**Discussion by xiegudong45** [NonceGeek/Move-Camp#250] [Q&A] [answered]
最近在回溯第一期课程，在研究aptos_framework::managed_coin包的源码时有些疑问：
1. Capabilities这个结构体为什么不能在coin包中定义，而是要在managed_coin中定义？如果是在coin中定义，那可以直接在coin::initialize中将Capabilities存放与账户中，看起来也比较简洁一些。是出于对coin包的扩展性而做考量吗？
2. mint方法中，mintCapability并没有被使用到，但为什么在调用的时候还是要求传入？
![image](https://user-images.githubusercontent.com/30063330/213868208-b9eda312-96e7-4048-8596-fbbadb17a1c6.png)

**Comment by qiwihui** [comment, marked as answer]
1. 是的，出于扩展性的考虑，coin中已经定义了 BurnCapability，FreezeCapability 和 MintCapability 这三种能力，但是在不用的场景中这三种能力不一定都用到，managed_coin 这个场景中用到了这三种能力。如果自己基于coin进行开发，不一定都需要这三种能力，比如我自己在写 WOOL 这个coin时就用 destroy_freeze_cap 方法去掉了 freeze 的能力： https://github.com/AptosWolfGame/wolf_game_contract/blob/master/Woolf/sources/wool.move#L17-L28

2. mintCapability 没有被使用但要求传入，这种其实要求调用 mint 方法的其他方法需要能够获取到 mintCapability ，也就是做了权限控制，只有有 mintCapability 的才能调用 mint 方法。可以参考 https://github.com/AptosWolfGame/wolf_game_contract/blob/master/Woolf/sources/wool.move#L52-L59
